### PR TITLE
C++ solution to Icy Perimeter

### PR DESCRIPTION
**Pull request by aldew5**

### CODE REVIEW

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        p ++;
+
+    ff(r + 1, c);
+    ff(r -1, c);
+    ff(r, c + 1);
+    ff(r, c -1);
+
+    return;
+
+}
```
Comment (HarryWangATX) [inline solutions/USACO feb19-perimeter-silver.mdx:79]: can u be consitent with code style here?

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+ifstream fin("perimeter.in");
+ofstream fout("perimeter.out");
+
+
+int n;
+char grid[1001][1001];
+bool visit[1001][1001];
+
+int a, p;
```
Comment (HarryWangATX) [inline solutions/USACO feb19-perimeter-silver.mdx:50]: just combine a,p with the first line.

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+#define check(x) cout<<(#x)<<": "<<x<<" " << endl;
+#define io ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
+#define ss second
+#define ff first
+#define pb push_back
+#define lb lower_bound
+#define ub upper_bound
+#define ld long double
+#define all(c) (c).begin(), (c).end()
+#define FOR(i, a, b) for (int i = a; i < b; i++)
```
Comment (HarryWangATX) [inline solutions/USACO feb19-perimeter-silver.mdx:31]: no loop macro

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                // test if a new max area was found
+                if (a >= mxa){
+                    // update maxp and maxa
+                    if (mxp != 0 && mxa == a)
+                        mxp = min(p, mxp);
+                    else
+                        mxp = p;
+                    mxa = a;
+
+                }
```
Comment (HarryWangATX) [inline solutions/USACO feb19-perimeter-silver.mdx:106]: this part can be more streamlined check official analysis

Comment (HarryWangATX) [reply solutions/USACO feb19-perimeter-silver.mdx:106]: might require some code changes in previous sections as well

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    visit[r][c] = true;
+    int co = 0;
+    if (r + 1 < n && grid[r + 1][c] == '.' || r + 1 >= n)
+         p++;
+    if (r -1 >= 0 && grid[r -1][c] == '.' || r - 1 < 0)
+        p ++;;
+    if (c + 1 < n && grid[r][c + 1] == '.' || c + 1 >= n)
+        p ++;
+    if (c -1 >= 0 && grid[r][c-1] == '.' || c -1 < 0)
+        p ++;
```
Comment (HarryWangATX) [inline solutions/USACO feb19-perimeter-silver.mdx:70]: official sol has a rly good impl for this

**Review by HarryWangATX — CHANGES_REQUESTED** [review]
might have to change up program a bit.

**Comment by HarryWangATX** [comment]
this is rly simiar to offical sol 👀 🤷

**Comment by maggie-j-liu** [comment]
Please follow the steps [here](https://usaco.guide/general/contributing?lang=cpp#adding-a-solution), in particular, make sure to modify the problem constructor and only use the macros mentioned there.

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+<LanguageSection>
+<CPPSection>
+
+
+```cpp
+
+#include <bits/stdc++.h>
+
+#define check(x) cout<<(#x)<<": "<<x<<" " << endl;
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:22]: remove this.

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+#define ss second
+#define ff first
+#define pb push_back
+#define lb lower_bound
+#define ub upper_bound
+#define ld long double
+#define all(c) (c).begin(), (c).end()
+#define FOR(i, a, b) for (int i = a; i < b; i++)
+#define mp make_pair
+#define line cout << "--------------" << endl;
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:33]: remove this aw

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+typedef long long ll;
+typedef unsigned long long ull;
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:36]: remove this

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+int pct(int x) { return __builtin_popcount(x); }
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:38]: remove this

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    if (r + 1 < n && grid[r + 1][c] == '.' || r + 1 >= n)
+         p++;
+    if (r -1 >= 0 && grid[r -1][c] == '.' || r - 1 < 0)
+        p ++;;
+    if (c + 1 < n && grid[r][c + 1] == '.' || c + 1 >= n)
+        p ++;
+    if (c -1 >= 0 && grid[r][c-1] == '.' || c -1 < 0)
+        p ++;
+
+    ff(r + 1, c);
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:72]: `ff(r+1, c);`

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    ff(r -1, c);
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:73]: `ff(r-1, c);`

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    ff(r, c + 1);
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:74]: `ff(r, c+1);`

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    ff(r, c -1);
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:75]: `ff(r, c-1);`

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+// flood fill that updates perimeter and area for 
+// each new block
+void ff (int r, int c){
+    if (r >= n || c >= n || r < 0 || c < 0 || grid[r][c] == '.' )
+        return;
+
+    if (visit[r][c]){
+        return;
+    }
+    a ++;
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:60]: `a++;`

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    visit[r][c] = true;
+    int co = 0;
+    if (r + 1 < n && grid[r + 1][c] == '.' || r + 1 >= n)
+         p++;
+    if (r -1 >= 0 && grid[r -1][c] == '.' || r - 1 < 0)
+        p ++;;
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:66]: `p++;`

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    if (c + 1 < n && grid[r][c + 1] == '.' || c + 1 >= n)
+        p ++;
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:68]: `p++;`

File: solutions/USACO feb19-perimeter-silver.mdx (hunk 5 above)
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:70]: `p++;`

File: solutions/USACO feb19-perimeter-silver.mdx [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+        return;
+    }
+    a ++;
+    visit[r][c] = true;
+    int co = 0;
+    if (r + 1 < n && grid[r + 1][c] == '.' || r + 1 >= n)
+         p++;
+    if (r -1 >= 0 && grid[r -1][c] == '.' || r - 1 < 0)
+        p ++;;
+    if (c + 1 < n && grid[r][c + 1] == '.' || c + 1 >= n)
```
Comment (jessechoe10) [inline solutions/USACO feb19-perimeter-silver.mdx:67]: put parentheses around your conditions so people aren't confused by the order of operations...

**Review by jessechoe10 — CHANGES_REQUESTED** [review]

**Comment by jessechoe10** [comment]
I'm gonna try submitting my solution a bit later, so don't merge yet lol.

**Comment by maggie-j-liu** [comment]
Also, use tabs instead of spaces.

**Comment by jessechoe10** [comment]
I think we can close this for now. When you fix the changes we requested, then submit another PR...

**Pull request closed without merging**